web/src/flow/stages/authenticator_validate/AuthenticatorValidateStage.ts

```diff
@@ -231,22 +234,40 @@ export class AuthenticatorValidateStage
         if (!this.challenge) {
             return html`<ak-empty-state loading> </ak-empty-state>`;
         }
+
         // User only has a single device class, so we don't show a picker
         if (this.challenge?.deviceChallenges.length === 1) {
             this.selectedDeviceChallenge = this.challenge.deviceChallenges[0];
         }
+
         // TOTP is a bit special, assuming that TOTP is allowed from the backend,
         // and we have a pre-filled value from the password manager,
         // directly set the the TOTP device Challenge as active.
         const totpChallenge = this.challenge.deviceChallenges.find(
             (challenge) => challenge.deviceClass === DeviceClassesEnum.Totp,
         );
-        if (PasswordManagerPrefill.totp && totpChallenge) {
+        if (PasswordManagerPrefill.totp && totpChallenge && !this._initialized) {
             console.debug(
                 "authentik/stages/authenticator_validate: found prefill totp code, selecting totp challenge",
             );
+            this._initialized = true;
             this.selectedDeviceChallenge = totpChallenge;
         }
+
+        // Pick the last used validator, if it's not a recovery key
+        const lastUsedChallenge = this.challenge.deviceChallenges
+            .filter((deviceChallenge) => deviceChallenge.lastUsed)
+            .sort((a, b) => b.lastUsed!.valueOf() - a.lastUsed!.valueOf())[0];
+        if (
+            lastUsedChallenge &&
+            lastUsedChallenge.deviceClass !== DeviceClassesEnum.Static &&
+            !this._initialized
+        ) {
+            this._initialized = true;
```

Comment:
Ah, don't change `property()` or `state()` fields inside `render()`. 

The problem is that I don't quite understand what this code is for.  If you need to set `this.selectedDeviceChallenge`, do it in the `willUpdate()` phase of the render.  [Changes inside `willUpdate()` cannot trigger rerenders.](https://lit.dev/docs/components/lifecycle/#willupdate)

Reply:
Also, is there a reason you need to track the `this._initialized`  at all?  What happens if you don't have it?  Could you have the `selectedDeviceChallenge` itself be the sigil you need to know which one to highlight?

Reply:
Let's say the user's last used device is TOTP. Without `this._initialized`, the following could happen:

1. User arrives to Authenticator Validate stage.
2. TOTP is automatically selected and the user is presented with the TOTP validation screen.
3. User decides to use webauthn instead, so clicks "Return to device picker".
4. TOTP is automatically selected again.

Effectively, without some countermeasure, if a device is autoselected the user can't go back to the device picker. `this._initialized` is supposed to be that countermeasure.

I agree that it smells. There should be a better place for the autoselection logic to run exactly once, right after `this.challenge` is loaded. What's that place? It's probably described somewhere in the `lit` documentation, but I haven't found it yet. Do you have any suggestions?

Reply:
Ah, okay, that makes a lot of sense.  It _is_ annoying that Lit's lifecycle has a `firstUpdated()` but not a `firstWillupdate()`, which is more or less what you're simulating here.